shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -275,15 +277,17 @@ public void removeOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedList
 
   // Called by native to notify first Flutter frame rendered.
   @SuppressWarnings("unused")
+  @VisibleForTesting
   @UiThread
-  private void onFirstFrame() {
+  protected void onFirstFrame() {
     ensureRunningOnMainThread();
     if (renderSurface != null) {
       renderSurface.onFirstFrameRendered();
     }
     // TODO(mattcarroll): log dropped messages when in debug mode (https://github.com/flutter/flutter/issues/25391)
-
-    for (OnFirstFrameRenderedListener listener : firstFrameListeners) {
+    CopyOnWriteArrayList<OnFirstFrameRenderedListener> mutableListeners =
```

Comment:
This leads me to believe that this is not a test of `FlutterRenderer` but rather a test of `FlutterJNI`. I think something needs to change in this PR to achieve correctness in that regard. If you're truly unit testing `FlutterRenderer` then you shouldn't have to make this protected. If you're testing `FlutterJNI` then the test suite should be switched accordingly.

Reply:
I think there's a couple of layers of testings we'll need either way. 

- I think we'll need strict unit tests. Where we're testing classes. I added a FlutterJNITest for this one. 
- Some of the classes are just not part of the public API are not going to be used. We should have medium tests that tests things between an API surface and some sort of abstraction layer. In this case, I think making things that platform_view_android_jni.cc calls also test callable probably makes sense. 
- We need integration tests that also includes C++ stuff but since those are heavier, they won't be per feature.